masterpassword.php

```diff
@@ -0,0 +1,97 @@
+<?php
+// This file is part of Moodle - http://moodle.org/
+//
+// Moodle is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// Moodle is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with Moodle.  If not, see <http://www.gnu.org/licenses/>.
+
+/**
+ * Master Password Settings
+ *
+ * @package    auth_basic
+ * @copyright  2018 Nathan Nguyen <nathannguyen@catalyst-au.nete>
+ * @license    http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
+ */
+
+require_once(__DIR__.'/../../config.php');
+require_once($CFG->libdir.'/adminlib.php');
+
+$action = optional_param('action', 0, PARAM_TEXT);
+$password = optional_param('password', null, PARAM_TEXT);
+$whitelistips = optional_param('whitelistips', null, PARAM_TEXT);
+
+require_login();
+require_capability('moodle/site:config', context_system::instance());
+
+admin_externalpage_setup('auth_basic_masterpassword');
+$thispage = '/auth/basic/masterpassword.php';
+$PAGE->set_url(new moodle_url($thispage));
+
+if (!isset($CFG->auth_basic_enabled_master_password)) {
+    echo $OUTPUT->header();
+    echo $OUTPUT->heading(get_string('auth_basic_masterpassword', 'auth_basic'));
+    echo html_writer::tag('div', get_string('auth_basic_notenabled_masterpassword', 'auth_basic'));
+    echo $OUTPUT->footer();
+} else {
+    if (($action === 'regenerate' && confirm_sesskey()) || empty($password) ) {
+        $password = time().uniqid();
+    } else if ($action === 'save' && confirm_sesskey() && !empty($password) ) {
+        global $DB;
+        $record = new stdClass();
+        $record->password = $password;
+        $record->userid = $USER->id;
+        $record->enabled = 1;
+        $record->ips = $whitelistips;
```

Comment:
Ah sorry this is my fault I was explicit - the ip whitelist should be set on config.php so a site admin cannot work around this.

Reply:
Actually a better way would just the ipaddress whitelist a normal admin setting which is shared by all master passwords, and then this can be optionally forced in config.php as needed.

The simplest way to do this would be to add this to the 'normal' basic auth settings, and I'm wondering whether we should check that whitelist across the board, with a real basic auth when the master password is not in use. The use of basic for the crawler would not be impacted by this. And in most cases you do probably want some extra level of protection if this is ever internet facing, so I think this does make sense.